File: script/python/tutorials/fwd1.py
# This pytoast example solves the forward problem
# for a homogeneous 2D circular problem

# Import various modules
import os
import sys
import numpy as np
from numpy import matrix
from scipy import sparse
from scipy.sparse import linalg
from numpy.random import rand
import matplotlib.pyplot as plt

# PyToast environment
execfile(os.getenv("TOASTDIR") + "/ptoast_install.py")
import toast

# Set the file paths
meshdir = os.path.expandvars("$TOASTDIR/test/2D/meshes/")
meshfile = meshdir + "circle25_32.msh"
qmfile = meshdir + "circle25_32x32.qm"

# Load the mesh and source/detector specs
mesh = toast.Mesh(meshfile)
mesh.ReadQM(qmfile)
nlen = mesh.NodeCount()

# Homogeneous parameter distributions
mua = np.ones ((1,nlen)) * 0.025
mus = np.ones ((1,nlen)) * 2.0
ref = np.ones ((1,nlen)) * 1.4
freq = 100

# Set up the linear system
smat = mesh.Sysmat (mua, mus, ref, freq)
qvec = mesh.Qvec ()
mvec = mesh.Mvec ()

# Solve the linear system
nq = qvec.shape[1]
phi = np.empty(qvec.shape,dtype='complex128')
for q in range(nq):
    qq = qvec[:,q].todense()
    res = linalg.minres(smat,qq,tol=1e-12)
    phi[:,q] = res[0]

# Project to boundary
y = mvec.transpose() * phi
logy = np.log(y)

# Display as sinogram
plt.figure(1)
im = plt.imshow(logy.real,interpolation='none')
plt.title('log amplitude')
plt.xlabel('detector index')
plt.ylabel('source index')
plt.colorbar()
plt.draw()
#plt.show()

plt.figure(2)
im = plt.imshow(logy.imag,interpolation='none')
plt.title('phase')
plt.xlabel('detector index')
plt.ylabel('source index')
plt.colorbar()
plt.show()

File: examples/matlab/fwd1.m
% Generate frequency DOT boundary data for a simple 2D circle,
% given parameter distributions and source/detector layout


clear all
close all

% Load mesh and optode layout
meshdir = '../../test/2D/meshes/';
hmesh = toastMesh([meshdir 'ellips_tri10.msh']);
hmesh.ReadQM([meshdir 'circle25_32x32.qm']);

% Load parameter distributions
n = hmesh.NodeCount;
mua = toastNim([meshdir 'tgt_mua_ellips_tri10.nim']);
mus = toastNim([meshdir 'tgt_mus_ellips_tri10.nim']);
ref = ones(n,1)*1.4;
freq = 100;

% Construct source and measurement vectors
qvec = hmesh.Qvec('Neumann','Gaussian',2);
mvec = hmesh.Mvec('Gaussian',2,ref);

% Solve FEM linear system
smat = dotSysmat(hmesh,mua,mus,ref,freq);
phi = smat\qvec;
meas = log(mvec.' * phi);

% Plot the log amplitude and phase sinograms
figure;
subplot(1,2,1);
imagesc(real(meas));
axis equal tight
colorbar
xlabel('source #');
ylabel('detector #');
title('log amp.');

subplot(1,2,2);
imagesc(imag(meas));
axis equal tight
colorbar
xlabel('source #');
ylabel('detector #');
title('phase');
